docs/convert_meta.1
.TH CONVERT_META 1 "Jan 25, 2012"
.\" Please adjust this date whenever revising the manpage.
.SH NAME
convert_meta \- A conversion program from .png map tiles to a more efficient conglomerate format (.meta).
.SH SYNOPSIS
.B convert_meta 
.RI [ options ]
.br
.SH DESCRIPTION
This manual page documents briefly the
.B convert_meta
command.
.PP
.B convert_meta
converts individual .png map tiles into the more efficient .meta file format. The .meta
format stores a grid of 8x8 tiles in a single meta tile, reducing the number of files by
a factor of 64.
.PP
.SH OPTIONS
This program follows the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
.TP
\fB\-t\fR|\-\-tile-dir
Specify the base directory where the tiles are stored. The default is /var/lib/mod_tile
.TP
\fB\-m\fR|\-\-map
Specify the map style to convert. This specifies the full path to the files and is relative to the tile directory.
The default is "default" and results in a full path of /var/lib/mod_tile/default to search for the tiles to convert
.TP
\fB\-u\fR|\-\-unpack
Unpack the .meta files back to PNGs
.TP
\fB\-z\fR|\-\-min-zoom
Specify the smallest zoom level to process. The default is 0.
.TP
\fB\-Z\fR|\-\-max-zoom
Specify the largest zoom level up to which to  process tiles. The default is 18.
.PP
.SH AUTHOR
convert_meta was written by Jon Burgess and other
OpenStreetMap project members.
.PP
This manual page was written by OpenStreetMap authors.


debian/changelog
libapache2-mod-tile (0.4-12~precise2) precise; urgency=low
  * fixed renderd.conf to use mapnik 2.0 plugins dir

 -- Kai Krueger <kakrueger@gmail.com>  Mon, 27 Aug 2012 00:00:00 +0000

libapache2-mod-tile (0.4-12~precise1) precise; urgency=low
  * new upstream version - revision 28638
    - add a json-tile handler
    - a number of bug fixes

 -- Kai Krueger <kakrueger@gmail.com>  Sun, 26 Aug 2012 00:00:00 +0000

libapache2-mod-tile (0.4-11~precise2) precise; urgency=low

  * correct renderd.conf
  * update upstream

 -- Kai Krueger <kakrueger@gmail.com>  Thu, 03 May 2012 00:00:00 +0000

libapache2-mod-tile (0.4-10~oneiric1) oneiric; urgency=low

  * support extensions other than png
  * autoconf'ify build system

 -- Kai Krueger <kakrueger@gmail.com>  Sun, 15 Jan 2012 00:00:00 +0000

libapache2-mod-tile (0.4-9~oneiric2) oneiric; urgency=low

  * added a slippymap
  * added scripts to update a tileserver

 -- Kai Krueger <kakrueger@gmail.com>  Sun, 09 Oct 2011 00:00:00 +0000

libapache2-mod-tile (0.4-8~natty1) natty; urgency=low

  * new upstream version
  * fixed debian package scripts

 -- Kai Krueger <kakrueger@gmail.com>  Thu, 06 Oct 2011 00:00:00 +0000

libapache2-mod-tile (0.3) jaunty; urgency=low

  * new version of render_list with proper multithreading
  * bugfixes
  * new render_expired

 -- Frederik Ramm <frederik@remote.org>  Tue, 24 Nov 2009 00:09:17 +0100

libapache2-mod-tile (0.2) jaunty; urgency=low

  * now also installs munin stuff 

 -- Frederik Ramm <frederik@remote.org>  Tue, 20 Oct 2009 19:52:04 +0200

libapache2-mod-tile (0.1) unstable; urgency=low

  * Initial Debian packaging for mod_tile and renderd

 -- Frederik Ramm <frederik@remote.org>  Thu, 04 Jun 2009 13:33:42 +0200


convert_meta.c
/*
Copyright © 2013 mod_tile contributors

This file is part of mod_tile.

mod_tile is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation, either version 2 of the License, or (at your
option) any later version.

mod_tile is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with mod_tile.  If not, see <http://www.gnu.org/licenses/>.
*/

#define _GNU_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/un.h>
#include <poll.h>
#include <errno.h>
#include <math.h>
#include <getopt.h>
#include <time.h>
#include <sys/types.h>
#include <dirent.h>
#include <assert.h>
#include <fcntl.h>
#include <limits.h>
#include <string.h>
#include "render_config.h"
#include "dir_utils.h"
#include "store.h"

char *tile_dir = HASH_PATH;

#ifndef METATILE
#warning("convert_meta not implemented for non-metatile mode. Feel free to submit fix")
int main(int argc, char **argv)
{
    fprintf(stderr, "convert_meta not implemented for non-metatile mode. Feel free to submit fix!\n");
    return -1;
}
#else

static int minZoom = 0;
static int maxZoom = MAX_ZOOM; 
static int verbose = 0;
static int num_render = 0, num_all = 0;
static struct timeval start, end;
static int unpack;

void display_rate(struct timeval start, struct timeval end, int num) 
{
    int d_s, d_us;
    float sec;

    d_s  = end.tv_sec  - start.tv_sec;
    d_us = end.tv_usec - start.tv_usec;

    sec = d_s + d_us / 1000000.0;

    printf("Converted %d tiles in %.2f seconds (%.2f tiles/s)\n", num, sec, num / sec);
    fflush(NULL);
}

static void descend(const char *search)
{
    DIR *tiles = opendir(search);
    struct dirent *entry;
    char path[PATH_MAX];

    if (!tiles) {
        //fprintf(stderr, "Unable to open directory: %s\n", search);
        return;
    }

    while ((entry = readdir(tiles))) {
        struct stat b;
        char *p;

        //check_load();

        if (!strcmp(entry->d_name, ".") || !strcmp(entry->d_name, ".."))
            continue;
        snprintf(path, sizeof(path), "%s/%s", search, entry->d_name);
        if (stat(path, &b))
            continue;
        if (S_ISDIR(b.st_mode)) {
            descend(path);
            continue;
        }
        p = strrchr(path, '.');
        if (p) {
            if (unpack) {
                if (!strcmp(p, ".meta")) 
                    process_unpack(tile_dir, path);
            } else {
                if (!strcmp(p, ".png")) 
                  process_pack(tile_dir, path);
            }
        }
    }
    closedir(tiles);
}


int main(int argc, char **argv)
{
    int z, c;
    const char *map = "default";

    while (1) {
        int option_index = 0;
        static struct option long_options[] = {
            {"map", 1, 0, 'm'},
            {"min-zoom", 1, 0, 'z'},
            {"max-zoom", 1, 0, 'Z'},
            {"unpack", 0, 0, 'u'},
            {"tile-dir", 1, 0, 't'},
            {"verbose", 0, 0, 'v'},
            {"help", 0, 0, 'h'},
            {0, 0, 0, 0}
        };

        c = getopt_long(argc, argv, "uhvz:Z:m:t:", long_options, &option_index);
        if (c == -1)
            break;

        switch (c) {
            case 'z':
                minZoom=atoi(optarg);
                if (minZoom < 0 || minZoom > MAX_ZOOM) { 
                    fprintf(stderr, "Invalid minimum zoom selected, must be between 0 and %d\n", MAX_ZOOM);
                    return 1;
                }
                break;
            case 'Z':
                maxZoom=atoi(optarg);
                if (maxZoom < 0 || maxZoom > MAX_ZOOM) {
                    fprintf(stderr, "Invalid maximum zoom selected, must be between 0 and %d\n", MAX_ZOOM);
                    return 1;
                }
                break;
            case 'm':
                map=strdup(optarg);
                break;
            case 't':
                tile_dir=strdup(optarg);
                break;
            case 'u':
                unpack=1;
                break;
            case 'v':
                verbose=1;
                break;
            case 'h':
                fprintf(stderr, "Usage: convert_meta [OPTION] ...\n");
                fprintf(stderr, "Convert the rendered PNGs into the more efficient .meta format\n");
                fprintf(stderr, "  -m, --map       convert tiles in this map (default is 'default')\n");
                fprintf(stderr, "  -t, --tile-dir  tile cache directory (default is '" HASH_PATH "')\n");
                fprintf(stderr, "  -u, --unpack    unpack the .meta files back to PNGs\n");
                fprintf(stderr, "  -z, --min-zoom  only process tiles greater or equal to this zoom level (default is 0)\n");
                fprintf(stderr, "  -Z, --max-zoom  only process tiles less than or equal to this zoom level (default is %d)\n", MAX_ZOOM);
                return -1;
            default:
                fprintf(stderr, "unhandled char '%c'\n", c);
                break;
        }
    }

    if (maxZoom < minZoom) {
        fprintf(stderr, "Invalid zoom range, max zoom must be greater or equal to minimum zoom\n");
        return 1;
    }

    fprintf(stderr, "Converting tiles in map %s\n", map);

    gettimeofday(&start, NULL);

    for (z=minZoom; z<=maxZoom; z++) {
        char path[PATH_MAX];
        snprintf(path, PATH_MAX, "%s/%s/%d", tile_dir, map, z);
        descend(path);
    }

    gettimeofday(&end, NULL);
    printf("\nTotal for all tiles converted\n");
    printf("Meta tiles converted: ");
    display_rate(start, end, num_render);
    printf("Total tiles converted: ");
    display_rate(start, end, num_render * METATILE * METATILE);
    printf("Total tiles handled: ");
    display_rate(start, end, num_all);

    return 0;
}
#endif
